langchain/evaluation/run_evaluators/config.py

```diff
@@ -0,0 +1,219 @@
+"""Configuration for run evaluators."""
+
+from typing import Any, Dict, List, Optional, Union
+
+from langsmith import RunEvaluator
+from pydantic import BaseModel, Field
+
+from langchain.embeddings.base import Embeddings
+from langchain.evaluation.criteria.eval_chain import CRITERIA_TYPE
+from langchain.evaluation.embedding_distance.base import EmbeddingDistance
+from langchain.evaluation.schema import EvaluatorType, StringEvaluator
+from langchain.evaluation.string_distance.base import StringDistance
+from langchain.schema.language_model import BaseLanguageModel
+from langchain.schema.prompt_template import BasePromptTemplate
+
+
+class EvalConfig(BaseModel):
+    """Configuration for a given run evaluator.
+
+    Parameters
+    ----------
+    evaluator_type : EvaluatorType
+        The type of evaluator to use.
+
+    Methods
+    -------
+    get_kwargs()
+        Get the keyword arguments for the evaluator configuration.
+
+    """
+
+    evaluator_type: EvaluatorType
+
+    def get_kwargs(self) -> Dict[str, Any]:
+        """Get the keyword arguments for the load_evaluator call.
+
+        Returns
+        -------
+        Dict[str, Any]
+            The keyword arguments for the load_evaluator call.
+
+        """
+        return self.dict(exclude={"evaluator_type"}, exclude_none=True)
+
+
+class RunEvalConfig(BaseModel):
+    """Configuration for a run evaluation.
+
+    Parameters
+    ----------
+    evaluators : List[Union[EvaluatorType, EvalConfig]]
+        Configurations for which evaluators to apply to the dataset run.
+        Each can be an evaluator type (e.g., "qa") or a configuration for a
+        given evaluator.
+
+    custom_evaluators : Optional[List[Union[RunEvaluator, StringEvaluator]]]
+        Custom evaluators to apply to the dataset run.
+
+    reference_key : Optional[str]
+        The key in the dataset run to use as the reference string.
+        If not provided, it will be inferred automatically.
+
+    prediction_key : Optional[str]
+        The key from the traced run's outputs dictionary to use to
+        represent the prediction. If not provided, it will be inferred
+        automatically.
+
+    input_key : Optional[str]
+        The key from the traced run's inputs dictionary to use to represent the
+        input. If not provided, it will be inferred automatically.
+
+    eval_llm : Optional[BaseLanguageModel]
+        The language model to pass to any evaluators that use a language model.
+    """
+
+    evaluators: List[Union[EvaluatorType, EvalConfig]] = Field(default_factory=list)
+    """Configurations for which evaluators to apply to the dataset run.
+    Each can be an evaluator type (e.g., "qa") or a configuration for a
+    given evaluator."""
+    custom_evaluators: Optional[List[Union[RunEvaluator, StringEvaluator]]] = None
```

Comment:
why do we allow both types here?

Reply:
I'm not sure what benefit we'd get from separating them out since it seems unlikely we'll see objects that inherit from both classes, but this lets people use custom run evaluators or evaluators that implement the langchain OSS interface